test/declarative_test_fwk/helper.py

```diff
@@ -0,0 +1,173 @@
+    Keep its methods stateless. Methods are on instance to allow specialization.
+    """
+    DEFAULT_WAIT_SECS = 30
+
+    def create(self, input_data: dict, input_replacements: dict = {}):
+        """
+        Creates custom resource
+        :param input_data: resource details
+        :param input_replacements: input replacements
+        :return: k8s.CustomResourceReference, created custom resource
+        """
+        reference = self.custom_resource_reference(input_data, input_replacements)
+        _ = k8s.create_custom_resource(reference, input_data)
+        resource = k8s.wait_resource_consumed_by_controller(reference, wait_periods=10)
+        assert resource is not None
+        return (reference, resource)
+
+    def patch(self, input_data: dict, input_replacements: dict = {}):
+        """
+        Patches custom resource
+        :param input_data: resource patch details
+        :param input_replacements: input replacements
+        :return: k8s.CustomResourceReference, created custom resource
+        """
+        reference = self.custom_resource_reference(input_data, input_replacements)
+        _ = k8s.patch_custom_resource(reference, input_data)
+        sleep(self.DEFAULT_WAIT_SECS)  # required as controller has likely not placed the resource in modifying
+        resource = k8s.wait_resource_consumed_by_controller(reference, wait_periods=10)
+        assert resource is not None
+        return (reference, resource)
+
+    def delete(self, reference: k8s.CustomResourceReference):
+        """
+        Deletes custom resource and waits for delete completion
+        :param reference: resource reference
+        :return: None
+        """
+        resource = k8s.get_resource(reference)
+        if not resource:
+            logging.warning(f"ResourceReference {reference} not found. Not invoking k8s delete api.")
+            return
+
+        k8s.delete_custom_resource(reference)
+        sleep(self.DEFAULT_WAIT_SECS)
+        self.wait_for_delete(reference) # throws exception if wait fails
+
+    def assert_expectations(self, verb: str, input_data: dict, expectations: dict, reference: k8s.CustomResourceReference):
+        """
+        Asserts custom resource reference against supplied expectations
+        :param verb: expectations after performing the verb (apply, patch, delete)
+        :param input_data: input data to verb
+        :param expectations: expectations to assert
+        :param reference: custom resource reference
+        :return: None
+        """
+        # condition assertion contains wait logic
+        self._assert_conditions(expectations, reference)
+
+        # conditions expectations met, now check current resource against expectations
+        resource = k8s.get_resource(reference)
+        self.assert_items(expectations.get("status"), resource.get("status"))
+
+        # self._assert_state(expectations.get("spec"), resource)  # uncomment to support spec assertions
```

Comment:
This is kinda funky because you are reading the CR from the Kubernetes API server multiple times (hidden to the user) but the test scenario acts like what is being asserted is simply the returned CR from the Create or Patch operation.

I feel like a more robust solution, if slightly more verbose, would be to have an explicit `wait` part of the `Step` structure that occurs before the `expect` stanza.

So, instead of this:

```yaml
steps:
  - id: "create_CMD_replication_group"
    description: "Initial config"
    create:
      spec:
        engine: redis
        replicationGroupID: test$RANDOM_SUFFIX
        replicationGroupDescription: Basic create and update of CMD RG
        cacheNodeType: cache.t3.micro
        numNodeGroups: 1
      expect:
        status:
          conditions:
            ACK.ResourceSynced: "True"
          status: "available"
```

You would do something like this:

```yaml
steps:
  - id: "create_CMD_replication_group"
    description: "Initial config"
    create:
      spec:
        engine: redis
        replicationGroupID: test$RANDOM_SUFFIX
        replicationGroupDescription: Basic create and update of CMD RG
        cacheNodeType: cache.t3.micro
        numNodeGroups: 1
    wait:
      status:
         conditions:
           ACK.ResourceSynced:
             value: "True"
             timeout: 30
    # These assertions are made on the resulting CR from a Kuberetes Describe operation
    # after all waits have successfully completed
    expect:
      status: "available"
```

That way, it's more explicit that there are actually distinct reads to the Kubernetes API server, with one set of reads having a retry/wait semantic and the other not.

Reply:
Updated code.
**`wait` can be specified as:**

1. Wait for 5 minutes
```
    wait: 5
```
2. Wait on condition expected value - waits for default timeout (30 minutes)
```
    wait:
      status:
        conditions:
          ACK.Terminal: "True"
```
I kept the `wait` details open ended, so that support to wait on other properties of custom resource can be be added later.

3. Wait on specific condition details with optional `timeout`:
```
    wait:
      status:
        conditions:
          ACK.ResourceSynced:
            status: "True"
            timeout: 30     # timeout in minutes, if not specified defaults to 30 minutes
```

**Assert on condition `message`:**

Conditions can be specified with expected `message` (both in `wait` as well as `expect`):
```
    status:
        conditions:
          ACK.Terminal: "True"
          ACK.ResourceSynced:
            status: "False"
            message: "Expected message ..."

```

Reply:
>     1. Wait for 5 minutes

Seconds, not minutes :)